File: bench/perl/src/ini2json.pl
#!/usr/bin/env perl
use 5.012;
use warnings;
use Time::HiRes qw/time/;

# Perl version of the INI 2 JSON conversion.
# Quite differently structured from the Kleenex version.

my $pre_compile = time;

my $commentRx  = qr/^\s*;|^\s*$/;
my $headingRx  = qr/^\s*\[([^\n\]]*)\]\s*$/;
my $keyValueRx = qr/^\s*([^;\s=\[]*)\s*=\s*([^\n]*?)\s*$/;
my $quotedRx   = qr/^".*"$/;
my $bslashRx   = qr/\\/;
my $quoteRx    = qr/"/;

# Start timing
my $start = time;

print "{\n";

my $firstSection = 1;
my $firstKeyValuePair = 1;
my $ind = "    ";

my $i = 0;
while (my $line = <>) {
    $i ++;
    next if ($line =~ $commentRx);

    if (my ($heading) = $line =~ $headingRx) {
        print "\n$ind},\n" unless $firstSection;
        $firstSection = 0;
        $firstKeyValuePair = 1;
        print "$ind\"$heading\": {";
        next;
    }

    die("Needs to start with a section before keys (line $i)") if $firstSection;

    if (my ($key, $value) = $line =~ $keyValueRx) {
        unless ($value =~ $quotedRx) {
            $value =~ s/$bslashRx/\\\\/g;
            $value =~ s/$quoteRx/\\"/g;
            $value = qq{"$value"};
        }

        print "," unless ($firstKeyValuePair);
        $firstKeyValuePair = 0;
        print qq{\n$ind$ind"$key": $value};
    }
}

print "\n$ind}" unless $firstSection;
print "\n}\n";

# End timing
my $end = time;
my $elaps = int(($end - $start) * 1000);
my $elaps_compile = int(($start - $pre_compile) * 1000);

print STDERR "\ncompilation (ms): ${elaps_compile}\n";
print STDERR "matching (ms):    ${elaps}\n"

File: bench/python/src/ini2json.py
#!/usr/bin/env python
import sys
import ConfigParser
import datetime
import re

pre_compile = datetime.datetime.now()

parser = ConfigParser.RawConfigParser()
parser.optionxform = str

bslash = re.compile(r'\\')
quote  = re.compile('"')

# Start timing
start = datetime.datetime.now()

parser.readfp(sys.stdin)
sys.stdout.write("{\n")

first_section = True

for section in parser.sections():
    if not first_section:
        sys.stdout.write(",\n")
    else:
        first_section = False

    sys.stdout.write("    \"%s\": {\n" % section)

    first_keyvalue = True
    for (key, value) in parser.items(section):
        if not first_keyvalue:
            sys.stdout.write(",\n")
        else:
            first_keyvalue = False

        if value.startswith('"') and value.endswith('"'):
            value = value[1:-1]
        else:
            value = bslash.sub(r'\\\\', value)
            value = quote.sub(r'\"', value)

        sys.stdout.write("        \"%s\": \"%s\"" % (key, value))

    if not first_keyvalue:
        sys.stdout.write("\n")
    sys.stdout.write("    }")

sys.stdout.write("\n}\n")

# End timing
end = datetime.datetime.now()

# Elapsed time
elaps = end - start
elaps_compile = start - pre_compile
elaps_ms = elaps.seconds * 1000 + elaps.microseconds / 1000
elaps_compile_ms = elaps_compile.seconds * 1000 + elaps_compile.microseconds / 1000

sys.stderr.write("\ncompilation (ms): %s\n" % str(elaps_compile_ms))
sys.stderr.write("matching (ms):    %s\n" % str(elaps_ms))
